File: examples/c++/Truss1D2N.cpp
#include <iostream>
#include "nuto/math/FullMatrix.h"
#include "nuto/math/SparseMatrixCSRGeneral.h"
#include "nuto/math/SparseDirectSolverMUMPS.h"
#include "nuto/mechanics/structures/unstructured/Structure.h"
#include "nuto/mechanics/structures/StructureOutputBlockVector.h"
#include "nuto/mechanics/constitutive/laws/AdditiveInputExplicit.h"
#include "nuto/mechanics/dofSubMatrixStorage/BlockScalar.h"
#include "nuto/mechanics/MechanicsEnums.h"
#include "nuto/visualize/VisualizeEnum.h"

int main()
{
    // definitions
    double YoungsModulus = 20000.;
    double Area = 100. * 100.;
    double Length = 1000.;
    int NumElements = 10;
    double Force = 1.;
    bool EnableDisplacementControl = false;
    double BoundaryDisplacement = 0.1;

    // create one-dimensional structure
    NuTo::Structure structure(1);

    // create section
    auto Section1 = structure.SectionCreate("Truss");
    structure.SectionSetArea(Section1, Area);

    // create material law
    auto Material1 = structure.ConstitutiveLawCreate("Linear_Elastic_Engineering_Stress");
    structure.ConstitutiveLawSetParameterDouble(
            Material1, NuTo::Constitutive::eConstitutiveParameter::YOUNGS_MODULUS, YoungsModulus);

    // create nodes
    NuTo::FullVector<double, Eigen::Dynamic> nodeCoordinates(1);
    for (int node = 0; node < NumElements + 1; node++)
    {
        std::cout << "create node: " << node << " coordinates: " << node * Length / NumElements << std::endl;
        nodeCoordinates(0) = node * Length / NumElements;
        structure.NodeCreate(node, nodeCoordinates);
    }

    int InterpolationType = structure.InterpolationTypeCreate("Truss1D");
    structure.InterpolationTypeAdd(InterpolationType, NuTo::Node::eDof::COORDINATES,
            NuTo::Interpolation::eTypeOrder::EQUIDISTANT1);
    structure.InterpolationTypeAdd(InterpolationType, NuTo::Node::eDof::DISPLACEMENTS,
            NuTo::Interpolation::eTypeOrder::EQUIDISTANT1);

    // create elements
    NuTo::FullVector<int, Eigen::Dynamic> elementIncidence(2);
    for (int element = 0; element < NumElements; element++)
    {
        std::cout << "create element: " << element << " nodes: " << element << "," << element + 1 << std::endl;
        elementIncidence(0) = element;
        elementIncidence(1) = element + 1;
        structure.ElementCreate(InterpolationType, elementIncidence);
        structure.ElementSetSection(element, Section1);
        structure.ElementSetConstitutiveLaw(element, Material1);
    }

    structure.ElementTotalConvertToInterpolationType();

    // set boundary conditions and loads
    NuTo::FullVector<double, Eigen::Dynamic> direction(1);
    direction(0) = 1;
    structure.ConstraintLinearSetDisplacementNode(0, direction, 0.0);
    structure.SetNumLoadCases(1);
    if (EnableDisplacementControl)
    {
        std::cout << "Displacement control" << std::endl;
        structure.ConstraintLinearSetDisplacementNode(NumElements, direction, BoundaryDisplacement);
    }
    else
    {
        std::cout << "Load control" << std::endl;
        structure.LoadCreateNodeForce(0, NumElements, direction, Force);
    }

    // start analysis
    structure.SolveGlobalSystemStaticElastic(1);
    auto residual = structure.BuildGlobalInternalGradient() - structure.BuildGlobalExternalLoadVector(1);

    std::cout << "residual: " << residual.J.CalculateNormL2() << std::endl;

    // visualize results
    int visualizationGroup = structure.GroupCreate(NuTo::eGroupId::Elements);
    structure.GroupAddElementsTotal(visualizationGroup);

    structure.AddVisualizationComponent(visualizationGroup, NuTo::eVisualizeWhat::DISPLACEMENTS);
    structure.AddVisualizationComponent(visualizationGroup, NuTo::eVisualizeWhat::ENGINEERING_STRAIN);
    structure.AddVisualizationComponent(visualizationGroup, NuTo::eVisualizeWhat::ENGINEERING_STRESS);
    structure.ExportVtkDataFileElements("Truss1D2N.vtk");

    return 0;
}

File: examples/python/Truss1D2N.py
# -*- coding: utf-8 -*-

# load nuto package
import nuto
import numpy as np

# definitions
YoungsModulus = 20000.
Area = 100. * 100.
Length = 1000.
NumElements = 10
Force = 1.
EnableDisplacementControl = False
BoundaryDisplacement = 0.1

# create one-dimensional structure
myStructure = nuto.Structure(1)

# create section
Section1 = myStructure.SectionCreate("TRUSS")
myStructure.SectionSetArea(Section1, Area)

# create material law
Material1 = myStructure.ConstitutiveLawCreate("Linear_Elastic_Engineering_Stress")
myStructure.ConstitutiveLawSetParameterDouble(Material1,"Youngs_Modulus", YoungsModulus)

# create nodes
nodeCoordinates = np.zeros(1)
for node in range(0, NumElements + 1):
    print "create node: " + str(node) + " coordinates: " + str(node * Length/NumElements)
    nodeCoordinates[0] = node * Length/NumElements
    myStructure.NodeCreateDOFs(node, "displacements", nodeCoordinates)

#create interpolation type
myInterpolationType = myStructure.InterpolationTypeCreate("Truss1D")
myStructure.InterpolationTypeAdd(myInterpolationType, "coordinates", "equidistant1")
myStructure.InterpolationTypeAdd(myInterpolationType, "displacements", "equidistant1")

# create elements
elementIncidence = nuto.IntVector(2)
for element in range(0, NumElements):
    print "create element: " + str(element) + " nodes: " + str(element) + "," + str(element+1)
    elementIncidence[0] = element
    elementIncidence[1] = element + 1
    myStructure.ElementCreate(element, myInterpolationType, elementIncidence)
    myStructure.ElementSetSection(element,Section1)
    myStructure.ElementSetConstitutiveLaw(element,Material1)

# set boundary conditions and loads
direction = nuto.DoubleFullMatrix(1,1,(1,))
myStructure.ConstraintLinearSetDisplacementNode(0, direction, 0.0)
myStructure.SetNumLoadCases(1)
if EnableDisplacementControl:
    print "Displacement control"
    myStructure.ConstraintLinearSetDisplacementNode(NumElements, direction, BoundaryDisplacement)
else:
    print "Load control"
    myStructure.LoadCreateNodeForce(0,NumElements, direction, Force)

#build maximum independent sets
myStructure.CalculateMaximumIndependentSets()

# start analysis
# start analysis
myStructure.SolveGlobalSystemStaticElastic(0)
intGradient = myStructure.BuildGlobalInternalGradient()
extGradient = myStructure.BuildGlobalExternalLoadVector(0)
residual = intGradient.J.Get("Displacements") - extGradient.J.Get("Displacements")


print "residual: " + str(np.linalg.norm(residual))

## visualize results
visualizationGroup = myStructure.GroupCreate("Elements")
myStructure.GroupAddElementsTotal(visualizationGroup)

myStructure.AddVisualizationComponent(visualizationGroup, "Displacements")
myStructure.AddVisualizationComponent(visualizationGroup, "EngineeringStrain")
myStructure.AddVisualizationComponent(visualizationGroup, "EngineeringStress")
#myStructure.ExportVtkDataFileElements("Truss1D2N.vtk")
